pkg/osbuild/mkfs_stage.go

```diff
@@ -7,81 +7,96 @@ import (
 
-		stageDevices, lastName := getDevices(path, filename, true)
+			// The last device in the chain must be named "device", because that's
+			// the device that mkfs stages run on. See the stage schemas for
+			// reference.
+			lastDevice := stageDevices[lastName]
+			delete(stageDevices, lastName)
+			stageDevices["device"] = lastDevice
 
-		// The last device in the chain must be named "device", because that's the device that mkfs stages run on.
-		// See their schema for reference.
-		lastDevice := stageDevices[lastName]
-		delete(stageDevices, lastName)
-		stageDevices["device"] = lastDevice
-
-		fsSpec := mnt.GetFSSpec()
-		switch t {
-		case "xfs":
-			options := &MkfsXfsStageOptions{
-				UUID:  fsSpec.UUID,
-				Label: fsSpec.Label,
-			}
-			stage = NewMkfsXfsStage(options, stageDevices)
-		case "vfat":
-			options := &MkfsFATStageOptions{
-				VolID: strings.Replace(fsSpec.UUID, "-", "", -1),
-			}
-			stage = NewMkfsFATStage(options, stageDevices)
-		case "btrfs":
-			// the disk library allows only subvolumes as Mountable, so we need to find the underlying btrfs partition
-			// and mkfs it
-			btrfsPart := findBtrfsPartition(path)
-			if btrfsPart == nil {
-				panic(fmt.Sprintf("found btrfs subvolume without btrfs partition: %s", mnt.GetMountpoint()))
+			switch e.GetFSType() {
+			case "xfs":
+				options := &MkfsXfsStageOptions{
+					UUID:  e.UUID,
+					Label: e.Label,
+				}
+				stages = append(stages, NewMkfsXfsStage(options, stageDevices))
+			case "vfat":
+				options := &MkfsFATStageOptions{
+					VolID: strings.Replace(e.UUID, "-", "", -1),
+				}
+				stages = append(stages, NewMkfsFATStage(options, stageDevices))
+			case "ext4":
+				options := &MkfsExt4StageOptions{
+					UUID:  e.UUID,
+					Label: e.Label,
+				}
+				stages = append(stages, NewMkfsExt4Stage(options, stageDevices))
+			default:
+				panic(fmt.Sprintf("unknown fs type: %s", e.GetFSType()))
 			}
+		case *disk.Btrfs:
+			stageDevices, lastName := getDevices(path, filename, true)
 
-			// btrfs partitions can be shared between multiple subvolumes, so we need to make sure we only create
-			// one
-			if processedBtrfsPartitions[btrfsPart.UUID] {
-				return nil
-			}
-			processedBtrfsPartitions[btrfsPart.UUID] = true
+			// The last device in the chain must be named "device", because that's
+			// the device that mkfs stages run on. See the stage schemas for
+			// reference.
+			lastDevice := stageDevices[lastName]
+			delete(stageDevices, lastName)
+			stageDevices["device"] = lastDevice
 
 			options := &MkfsBtrfsStageOptions{
-				UUID:  btrfsPart.UUID,
-				Label: btrfsPart.Label,
+				UUID:  e.UUID,
+				Label: e.Label,
 			}
-			stage = NewMkfsBtrfsStage(options, stageDevices)
-		case "ext4":
-			options := &MkfsExt4StageOptions{
-				UUID:  fsSpec.UUID,
-				Label: fsSpec.Label,
+			stages = append(stages, NewMkfsBtrfsStage(options, stageDevices))
+			// Handle subvolumes here directly instead of collecting them in
```

Comment:
Maybe add a line about the /why/ here as well? I assume just because of simplify? Or is there a deeper reason? Maybe `// Handle subvolumes directly here instead of in their own case because we can reuse the parent easier this way` (I totally made this up :)

Reply:
Sure.